Testing: Upgrade Jest to v27

gziolo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
Depends on #36328.

See details about breaking changes introduced in Jest v 27:
https://jestjs.io/blog/2021/05/25/jest-27

There is some bigger issue with unit tests that import `webpack` to build test fixtures. It either doesn't run at all or it takes forever and tests timeout.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

gziolo committed:
Testing: Upgrade Jest to v27

gziolo commented on packages/jest-preset-default/jest-preset.js:
```diff
@@ -26,5 +27,5 @@ module.exports = {
 		'^.+\\.[jt]sx?$': require.resolve( 'babel-jest' ),
 	},
-	verbose: true,
+	verbose: false,
```
`false` is the default value so it can be safely removed. It only slows down test execution and it should never be enabled in the default preset.

gziolo replied:
Removed with https://github.com/WordPress/gutenberg/pull/33287/commits/cbf38a471eabe18f74fd7762b2088a4467fe9b2a

gziolo commented:
The only remaining blocker is failing tests for React Native code. There is still something wrong with the configuration.

simonhammes commented:
@gziolo I think I've managed to solve the issue of the JS unit tests timing out:

Apparently this has to do with a change in Jest concerning async hooks.

There are no timeouts after applying this patch (I know that you should generelly prefer async methods whenever IO is involved):

```patch
diff --git a/test/integration/full-content/full-content.test.js b/test/integration/full-content/full-content.test.js
index 64862d19f0..363169c857 100644
--- a/test/integration/full-content/full-content.test.js
+++ b/test/integration/full-content/full-content.test.js
@@ -58,8 +58,8 @@ function normalizeParsedBlocks( blocks ) {
 }
 
 describe( 'full post content fixture', () => {
-	beforeAll( async () => {
-		const blockMetadataFiles = await glob(
+	beforeAll( () => {
+		const blockMetadataFiles = glob.sync(
 			'packages/block-library/src/*/block.json'
 		);
 		const blockDefinitions = fromPairs(
```

simonhammes commented:
And most (if not all of them) of the native test errors should be gone once RN is updated to 0.66 (https://github.com/WordPress/gutenberg/pull/36328), since that would include @jest/create-cache-key-function@27 (see https://github.com/callstack/react-native-testing-library/issues/743)

gziolo commented:
> Apparently this has to do with a change in Jest concerning async hooks.
>
> There are no timeouts after applying this patch (I know that you should generelly prefer async methods whenever IO is involved)

Interesting discovery. Do you know what has changed in regards to async hooks? I would love to learn more.

simonhammes commented:
> > Apparently this has to do with a change in Jest concerning async hooks.
> > There are no timeouts after applying this patch (I know that you should generelly prefer async methods whenever IO is involved)
> 
> Interesting discovery. Do you know what has changed in regards to async hooks? I would love to learn more.

No, sorry, I did not investigate this any further.

EDIT: I guess I could try and find a minimal example which reproduces this bug and open an issue over at facebook/jest.

gziolo commented:
I rebased the branch with the most recent changes in `trunk`. There is now Jest `27.4.4` so I updated dependencies again. I also applied the fix that @simonhammes suggested in https://github.com/WordPress/gutenberg/pull/33287#issuecomment-983991349.

> And most (if not all of them) of the native test errors should be gone once RN is updated to 0.66 (#36328), since that would include @jest/create-cache-key-function@27 (see callstack/react-native-testing-library#743)

We are still waiting for React Native upgrade.

geriux commented:
> We are still waiting for React Native upgrade.

FYI The [upgrade](https://github.com/WordPress/gutenberg/pull/36328) was just merged 🎉

gziolo commented:
> > We are still waiting for React Native upgrade.
> 
> FYI The [upgrade](https://github.com/WordPress/gutenberg/pull/36328) was just merged 🎉

Awesome. I saw the commit earlier today. I will plan to bring this PR up to date this week. Thank you for making it happen 👏🏻

geriux commented:
> > > We are still waiting for React Native upgrade.
> > 
> > 
> > FYI The [upgrade](https://github.com/WordPress/gutenberg/pull/36328) was just merged 🎉
> 
> Awesome. I saw the commit earlier today. I will plan to bring this PR up to date this week. Thank you for making it happen 👏🏻

Cool! Kudos to the people involved with the upgrade @dcalhoun @twstokes @mchowning @hypest 👏 let us know if you need any further help 👍

gziolo commented:
I updated this branch with the latest changes in `trunk`. It looks like `npm run native test` is very slow now. It takes 2-3 minutes on CI for `trunk` (https://github.com/WordPress/gutenberg/runs/4699951957?check_suite_focus=true). While here it's over 20 minutes and it is still running. I would appreciate some help with debugging because it was working perfectly fine before the upgrade to RN 0.66 landed.

dcalhoun committed:
Fix native test timeouts caused by combining fake timers and setImmediate (#37715)

* Upgrade to @testing-library/react-native@8.0.0

* Clean up fake timer usage after tests

This may be unnecessary, but avoid potential issues where fake timers
are unexpectedly used and cause breakages in other tests.

* Enable combination of modern fake timers and waitFor

Previously, `waitFor` would timeout when `jest.useFakeTimers('modern')`
was enabled. The 'modern' version is now the default in Jest 27.

The Jest preset from `@testing-library/react-native` provides a
workaround for a larger issue in React Native and Jest that mutates the
global `Promise` object. https://git.io/JSDZI

* Remove global enabling of fake timers

Enabling fake timers can have negative consequences with `waitFor`, e.g.
causing unexpected timeouts. Enabling it globally is far-reaching and
this should likely be enabled within individual tests as needed.

* Replace jest-jasmine2 with jest-circus

The latter is considered the successor to the former. We seemingly do
not depend on anything explicitly provided by `jest-jasmine2` and should
likely move on from it.

* Switch testing environment from jsdom to node

Improves speed of test environment setup and fixes a timeout issue when
combining `waitFor` and `jest.useFakeTimers('modern')`. It is not yet
exactly pinpointed as to _why_ this fixes the timeout issue. It appears
to related to `setImmediate` and `setTimeout`.

* Remove setImmediate global for testing environment

This may be unnecessary if `testEnvironment: 'node'` is retained.
However, most tests are currently broken due to missing DOM APIs.

* Polyfill required DOM APIs for testing environment

Now that `testEnvironment: 'node'` is utilized for the testing
environment, we must mirror the app runtime and polyfill the necessary
DOM APIs used in the source.

The Enzyme configuration removed conflicts with the switch from
`testEnvironment: 'jsdom'` to `testEnvironment: 'node'`. Enzyme depends
upon `react-dom`, which introduces far more dependencies upon DOM APIs.
Currently, all Enzyme-related tests fail and need to be replaced with
`@testing-library/react-native`.

* Avoid import of react-dom within native file

Importing `react-dom` introduces additional dependencies upon the DOM
API and is incompatible with `testEnvironment: 'node'`. The `act`
utility is available from `@testing-library/react-native`.

* Explicitly toggle fake timers in tests

This may not be necessary, but may help avoid unexpected issues from
lingering fake timers, e.g. timeout errors while using `waitFor`.

* Reinstate legacy Jest timers

The Jest preset from `@testing-library/react-native` that fixed support
for "modern" timers by modifying the polyfilled the global `Promise`
resulted in new failures from within React Native itself. Specifically,
core React Native components rely upon `.done` from the `promise`
package. `.done` is a non-standard method that does not exist on the
global `Promise` used by `@testing-library/react-native`'s preset.

* Disable erroneously failing test

This previously passing test now fails after upgrading
`@testing-library/react-native` due to changes in the library. Setting
`pointerEvent` to "box-none" or "none" currently erroneously prevents
triggering other events unrelated to pressing on the element, e.g.
`onTouch*`, `onLayout`.

https://git.io/JSHZt

* Reinstate jest-jasmine2 to support done callback

The Jest team create jest-circus as the predecessor for jest-jasmine2.
The former does not support the `done` callback with async/await, and
would appear to have no plans to do so.

It would likely benefit us to refactor the one current test using
`done` away from it, and embrace `jest-circus` to maintain alignment
with Jest core.

* Refactor native unit tests away from done callback

The Jest team created `jest-circus` as the predecessor for
`jest-jasmine2`. The former does not support the `done` callback with
async/await, and would appear to have no plans to do so.

In order to embrace `jest-circus` and maintain alignment with Jest core,
the one test using `done` was refactored to avoid it

- https://git.io/JSHWU
- https://git.io/JSHWk

* Remove Enzyme from Editor tests

* Remove Enzyme from Paragraph tests

* Remove Enzyme from BlockMover tests

* Remove Enzyme from BlockEdit test

* Remove Enzyme from LinksUI test

* Remove Enzyme from ListEdit tests

* Remove Enzyme from Platform tests

* Remove Enzyme from BlockTypesTab tests

* Remove Enzyme from HTMLTextInput tests

* Remove Enzyme from ReusableBlockTab tests

* Remove Enzyme from MediaUpload tests

* Remove Enzyme from BlockMediaUpdateProgress tests

* Remove Enzyme from MediaUploadProgress tests

* Fix ReferencEerror in Inserter test

Usage of `react-test-renderer` caused the following error. Leveraging
`@testing-library/react-native` instead resolved it for an unknown
reason.

```
ReferenceError: You are trying to `import` a file after the Jest environment has been torn down. From packages/block-editor/src/components/inserter/test/index.native.js.

      468 |                     style: listStyle,
      469 |                     safeAreaBottomInset,
    > 470 |                     scrollEnabled,
          |                                             ^
      471 |                     automaticallyAdjustContentInsets: false,
      472 |             };
      473 |

      at Object.get PanResponder [as PanResponder] (node_modules/react-native/index.js:251:12)
      at BottomSheet.render (packages/components/src/mobile/bottom-sheet/index.native.js:470:39)
      at finishClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8459:31)
      at updateClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8409:24)
      at beginWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:9990:16)
```

* Fix ReferencEerror in Verse test

Usage of `react-test-renderer` caused the following error. Leveraging
`@testing-library/react-native` instead resolved it for an unknown
reason.

```
ReferenceError: You are trying to `import` a file after the Jest environment has been torn down. From packages/block-editor/src/components/inserter/test/index.native.js.

      468 |                     style: listStyle,
      469 |                     safeAreaBottomInset,
    > 470 |                     scrollEnabled,
          |                                             ^
      471 |                     automaticallyAdjustContentInsets: false,
      472 |             };
      473 |

      at Object.get PanResponder [as PanResponder] (node_modules/react-native/index.js:251:12)
      at BottomSheet.render (packages/components/src/mobile/bottom-sheet/index.native.js:470:39)
      at finishClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8459:31)
      at updateClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8409:24)
      at beginWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:9990:16)
```

* Fix ReferencEerror in Audio test

Usage of `react-test-renderer` caused the following error. Leveraging
`@testing-library/react-native` instead resolved it for an unknown
reason.

```
ReferenceError: You are trying to `import` a file after the Jest environment has been torn down. From packages/block-editor/src/components/inserter/test/index.native.js.

      468 |                     style: listStyle,
      469 |                     safeAreaBottomInset,
    > 470 |                     scrollEnabled,
          |                                             ^
      471 |                     automaticallyAdjustContentInsets: false,
      472 |             };
      473 |

      at Object.get PanResponder [as PanResponder] (node_modules/react-native/index.js:251:12)
      at BottomSheet.render (packages/components/src/mobile/bottom-sheet/index.native.js:470:39)
      at finishClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8459:31)
      at updateClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8409:24)
      at beginWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:9990:16)
```

* Fix ReferencEerror in File test

Usage of `react-test-renderer` caused the following error. Leveraging
`@testing-library/react-native` instead resolved it for an unknown
reason.

```
ReferenceError: You are trying to `import` a file after the Jest environment has been torn down. From packages/block-editor/src/components/inserter/test/index.native.js.

      468 |                     style: listStyle,
      469 |                     safeAreaBottomInset,
    > 470 |                     scrollEnabled,
          |                                             ^
      471 |                     automaticallyAdjustContentInsets: false,
      472 |             };
      473 |

      at Object.get PanResponder [as PanResponder] (node_modules/react-native/index.js:251:12)
      at BottomSheet.render (packages/components/src/mobile/bottom-sheet/index.native.js:470:39)
      at finishClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8459:31)
      at updateClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8409:24)
      at beginWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:9990:16)
```

* Fix ReferencEerror in Search test

Usage of `react-test-renderer` caused the following error. Leveraging
`@testing-library/react-native` instead resolved it for an unknown
reason.

```
ReferenceError: You are trying to `import` a file after the Jest environment has been torn down. From packages/block-editor/src/components/inserter/test/index.native.js.

      468 |                     style: listStyle,
      469 |                     safeAreaBottomInset,
    > 470 |                     scrollEnabled,
          |                                             ^
      471 |                     automaticallyAdjustContentInsets: false,
      472 |             };
      473 |

      at Object.get PanResponder [as PanResponder] (node_modules/react-native/index.js:251:12)
      at BottomSheet.render (packages/components/src/mobile/bottom-sheet/index.native.js:470:39)
      at finishClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8459:31)
      at updateClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8409:24)
      at beginWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:9990:16)
```

* Fix ReferencEerror in Missing test

Usage of `react-test-renderer` caused the following error. Leveraging
`@testing-library/react-native` instead resolved it for an unknown
reason.

```
ReferenceError: You are trying to `import` a file after the Jest environment has been torn down. From packages/block-editor/src/components/inserter/test/index.native.js.

      468 |                     style: listStyle,
      469 |                     safeAreaBottomInset,
    > 470 |                     scrollEnabled,
          |                                             ^
      471 |                     automaticallyAdjustContentInsets: false,
      472 |             };
      473 |

      at Object.get PanResponder [as PanResponder] (node_modules/react-native/index.js:251:12)
      at BottomSheet.render (packages/components/src/mobile/bottom-sheet/index.native.js:470:39)
      at finishClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8459:31)
      at updateClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8409:24)
      at beginWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:9990:16)
```

* Upgrade to @testing-library/react-native@9.0.0

* Add assertions to block type tab tests

Improve test clarity with explicit expect assertions.

Co-authored-by: Carlos Garcia <fluiddot@gmail.com>

* Remove unnecessary abstraction

The switch away from Enzyme reduced this abstraction to a single line,
so it provides less value now.

* Consistently assert media block update progress spinner removal

This increases consistency amongst the tests, as most already include
this assertion.

* Update MediaUpload test to select 'Choose from device' option

This selection better aligns with the test description.

* Remove duplicative matchMedia global definition

The test environment now imports the globals setup file used by the app
runtime. That file includes a `matchMedia` global definition as well.

* Add assertions to LinkSettings tests

Improve test clarity with explicit expect assertions.

* Removed shallow renderer usage in tests

Shallow rendering components is generally considered a non-optimal
approach to testing React components by the React community. This
replaces `shallow` with `render` to further test integration of the
subject components.

* Remove unused import

The `React` import was utilized by the now removed `shallow` render
implementation.

* Remove unnecessary top-level beforeAll usage

Jest runs each test file independently, so top-level code will not
impact other test files. Since the `(before|after)All` usage in code
changed is all top-level, and not within a `describe`, it is
superfluous.

Co-authored-by: Carlos Garcia <fluiddot@gmail.com>

dcalhoun commented:
@gziolo I am uncertain as to why the mobile tests are still failing after https://github.com/WordPress/gutenberg/pull/33287/commits/58725edcc6f2ffce4462ffe5fa9a5dd414b069da. They succeeded in #37715. 

However, I have seen the current `TypeError: Cannot use 'in' operator to search for 'WebkitAnimation' in undefined` error before. It previously originated from Enzyme forcing loading of `react-dom`, which fails to run after switching the test environment from `jsdom` to `node`. All usage of Enzyme was removed from `*.native.js` test files. I verified that to be the case in this branch for the currently failing files. 

I wonder if this may be a test environment caching issue? How might we ensure a fresh cache?

gziolo commented:
I rebased the branch with the latest changes from `trunk` hoping it resets the cache, but it looks like I see the same failures locally:

<img width="1126" alt="Screenshot 2022-01-12 at 15 35 24" src="https://user-images.githubusercontent.com/699132/149160462-eff11199-6b7c-410d-8158-1217dd939c8c.png">

Is it possible that some changes on `trunk` got applied in the meantime which require JSDom to pass? WebkitAnimation is definitely not something you would expect in React Native.

gziolo commented:
@dcalhoun, I updated a bunch of dev dependencies so maybe it's related to that. Anyway, it would be good to find out why the code would want to use `WebkitAnimation` in the first place. Maybe, it's something that needs an override for mobile apps.

dcalhoun commented:
@gziolo I opened https://github.com/WordPress/gutenberg/pull/37921 to address the current native test failures.

gziolo commented:
> @gziolo I opened #37921 to address the current native test failures.

It looks like it does the trick. I'm super excited about all the improvements coming with this PR to the native unit tests! I'd appreciate some sanity check for the web changes in this PR as well.

Separately from this PR we can look into:
- switching to `node` as the default test env for web unit tests instead of `jsdom` that I set for backward compatibility
- switching to the new test runner in native e2e tests

dcalhoun committed:
Fix erroneous react-dom usage in native tests (#37921)

* Reinstate react-platform.native.js file

This file was removed in the following commit, as it was no longer
strictly necessary. https://github.com/WordPress/gutenberg/commit/b7b62d2d3db4395ce5d2ad42bd1fc0f1c3c2f12d#diff-f02069349f238fb47a268bb7fcc03c9768331db18ead0e28d8ecad7bbc05037c

However, the removal led to Jest loading the DOM-specific version of the
file when mocking Gutenberg modules that depended upon
`react-platform.js`. Jest would seemingly load `react-dom` when
attempting to auto-mock a module, e.g. `jest.mock( '@wordpress/data/src/components/use-select' );`.

The loading of `react-dom` resulted in the following error:

```
TypeError: Cannot use 'in' operator to search for 'WebkitAnimation' in undefined

  at getVendorPrefixedEventName (node_modules/react-dom/cjs/react-dom.development.js:5011:58)
  at node_modules/react-dom/cjs/react-dom.development.js:5019:21
  at Object.<anonymous> (node_modules/react-dom/cjs/react-dom.development.js:26261:5)
  at Object.<anonymous> (node_modules/react-dom/index.js:37:20)
```

Reinstating `react-platform.native.js` addresses this issue by ensuring
that Jest does not encounter an import of `react-dom`.

* Fix false-positive ReactNativeEditor test

This test failed when run in isolation. It would appear it was dependent
upon the `jest.mock('../setup')` found in sibling tests.

After defining a mock for this specific test, it was discovered that the
assertion did not await the asynchronous query found within the test.
The assertion resulted in a false-positive as the returned query
`Promise` technically matches `toBeDefined`. `async`/`await` was added
to ensure the test assertion awaits the query result, as well as fixes
the following related log warning.

```
A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks. Active timers can also cause this, ensure that .unref() was called on them.
```

gziolo commented:
@dcalhoun and @fluiddot – let's see if we finally get green light from CI 😄

dcalhoun commented:
> switching to `node` as the default test env for web unit tests instead of `jsdom` that I set for backward compatibility

Is the target with this idea to avoid the significant performance overhead incurred by `jsdom`? Given the web tests are so DOM-centric, leveraging `jsdom` does make some sense. Would the alternative plan be to maintain mocks of the specific DOM APIs required? 

> switching to the new test runner in native e2e tests

Agreed. I considered doing this in https://github.com/WordPress/gutenberg/pull/37715, but opted to postpone that work to keep the PR focused purely on fixing the failing tests. Hopefully we can update the e2e tests soon as well.

gziolo commented:
> > switching to node as the default test env for web unit tests instead of jsdom that I set for backward compatibility
>
> Is the target with this idea to avoid the significant performance overhead incurred by jsdom? Given the web tests are so DOM-centric, leveraging jsdom does make some sense. Would the alternative plan be to maintain mocks of the specific DOM APIs required?

It's not only performance but making it explicit where a test needs DOM. Overall, it would lead to a better design that works better with server-side rendering (https://github.com/WordPress/gutenberg/issues/17273) and React Native. You can always explicitly set [`testEnvironment`](https://jestjs.io/docs/configuration#testenvironment-string) to `jsdom` for individual tests with:

```js
/**
 * @jest-environment jsdom
 */
```

gziolo commented:
I rebased this branch with `trunk` again. I see that there is a new test failing. It was introduced by @geriux in https://github.com/WordPress/gutenberg/pull/37951. @dcalhoun or @fluiddot, can you help to fix it? I'm able to reproduce it locally, but I can't figure out why it complains because `act` is used internally in `initializeEditor`:

<img width="872" alt="Screenshot 2022-01-18 at 14 11 20" src="https://user-images.githubusercontent.com/699132/149943700-d6d9b35f-226b-4ba7-bd0d-0b5d5c16edfb.png">

Once we have this test sorted out, I will merge this PR to avoid the hassle of bringing the branch up to date. I think it had enough testing from a few people already.

fluiddot commented:
> I rebased this branch with `trunk` again. I see that there is a new test failing. It was introduced by @geriux in #37951. @dcalhoun or @fluiddot, can you help to fix it?

Thanks for raising this issue @gziolo 🙇 , I'll take a look at this as soon as possible 💨 .

fluiddot committed:
[RNMobile] Fix `act` warnings that might be triggered after test finishes (#38052)

* Fix act warnings after test finishes

This is a workaround for addressing the act warnings that might caused by store updates executed after the test finishes.

* Update comment of setImmediate workaround

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

* Unmount editor when test finish

* Update initialize editor helper

The function is now synchronous as we don't have to wait for any element.

* Update tests that use initialize editor

* Fix act warnings from store resolvers with fake timers (#38077)

* Leverage fake timers to resolve store resolvers

During editor initialization, asynchronous store resolvers rely upon
`setTimeout` to run at the end of the current JavaScript block
execution. In order to prevent "act" warnings triggered by updates to
the React tree, we leverage fake timers to manually tick and await the
resolution of the current block execution before proceeding.

* Update RichText test `initializeEditor` usage

The refactor of `initializeEditor` to leverage fake timers means this
test no longer needs to `await` asynchronous work or manuall `unmount`
its subject component.

* Refactor usage of `initializeEditor` in tests

Now that `initializeEditor` leverages fake timers to resolve
asynchronous store resolvers, the tests no longer need to await a
resolution of `initializeEditor`.

* Fix `act` warnings in Image edit test

Retrieving text from the clipboard is an asynchronous action. The
component updated React state once the clipboard resolved. This resulted
in a state update triggering an `act` warning. Because there is no
visible change to the rendered output, e.g. updated text or UI, we must
await the resolution of the clipboard promise itself.

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

gziolo commented:
I hope to merge this PR as first thing in the morning. Let's hope nothing new doesn't pop up after I rebase with `trunk` 🤞🏻

dcalhoun commented on packages/url/src/test/index.native.js:
```diff
@@ -35,6 +35,18 @@ jest.mock( './fixtures/wpt-data.json', () => {
 		'file://­/p',
 		'file://%C2%AD/p',
 		'file://xn--/p',
+		'http://a.b.c.xn--pokxncvks',
+		'http://10.0.0.xn--pokxncvks',
+		'foo://ho|st/',
+		'http://ho%3Cst/',
+		'http://ho%3Est/',
+		'http://ho%7Cst/',
+		'file://%43%7C',
+		'file://%43|',
+		'file://C%7C',
+		'file://%43%7C/',
+		'https://%43%7C/',
+		'asdf://%43|/',
```
What is the purpose of these additions?

gziolo replied:
The implementation of URL polyfill for React Native doesn’t fully align with the one used in the browser. The text fixture we downloaded passes all checks for web but we need to list exceptions for native implementation. If the list of exceptions grows too much we might need to revisit this approach.

dcalhoun reviewed (approved):
🚀 LGTM. Thank you for the hard work to land this upgrade. Some really great improvements in this PR. 👏🏻

gziolo commented:
Great team work. Thank you all for all the help.
